fdbserver/DataDistribution.actor.cpp

```diff
@@ -396,6 +425,153 @@ struct DataDistributor : NonCopyable, ReferenceCounted<DataDistributor> {
+				if (auditState.getPhase() == AuditPhase::Error) {
+					continue;
+				} else if (auditState.getPhase() == AuditPhase::Failed) {
+					if (numFinishAuditsCleared < numFinishAuditsToClear) {
+						// Clear both audit metadata and corresponding progress metadata
+						self->addActor.send(clearAuditMetadata(self->txnProcessor->context(),
+						                                       auditState.getType(),
+						                                       auditState.id,
+						                                       /*clearProgressMetadata=*/true));
+						numFinishAuditsCleared++;
+					}
+					continue;
+				} else if (auditState.getPhase() == AuditPhase::Complete) {
+					if (numFinishAuditsCleared < numFinishAuditsToClear) {
+						// Clear audit metadata only
+						// No need to clear the corresponding progress metadata
+						// since it has been cleared for Complete audits
+						self->addActor.send(clearAuditMetadata(self->txnProcessor->context(),
+						                                       auditState.getType(),
+						                                       auditState.id,
+						                                       /*clearProgressMetadata=*/false));
+						numFinishAuditsCleared++;
+					}
+					continue;
+				}
+				ASSERT(auditState.getPhase() == AuditPhase::Running);
+				if (self->audits.contains(auditState.getType()) &&
+				    self->audits[auditState.getType()].contains(auditState.id)) {
+					// It is possible that the current DD is running this audit
+					// Suppose DDinit re-runs right after a new audit is persisted
+					// For this case, auditResume sees the new audit and resumes it
+					// At this point, the new audit is in the audit map
+					// Meanwhile, existing DDAuditStorage retries (since DD does not restart)
+					// Then, existing DDAuditStorageStart is conflict with auditResume
+					// auditResume cannot see the new audit in the map
+					continue;
+				}
+				TraceEvent(SevDebug, "AuditStorageResume", self->ddId)
+				    .detail("AuditID", auditState.id)
+				    .detail("AuditType", auditState.getType())
+				    .detail("AuditState", auditState.toString())
+				    .detail("NumFinishAuditsCleared", numFinishAuditsCleared)
+				    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+				runAuditStorage(self, auditState, 0, "ResumeAudit");
+			}
+		}
+		self->auditInitialized.send(Void());
+		TraceEvent(SevInfo, "AuditStorageResumeDone", self->ddId);
+		return;
+	}
+
+	ACTOR static Future<Void> initAuditStorage(Reference<DataDistributor> self, int selfGeneration) {
+		MoveKeyLockInfo lockInfo;
+		lockInfo.myOwner = self->lock.myOwner;
+		lockInfo.prevOwner = self->lock.prevOwner;
+		lockInfo.prevWrite = self->lock.prevWrite;
+		std::vector<AuditStorageState> auditStates = wait(loadAndUpdateAuditMetadataWithNewDDId(
+		    self->txnProcessor->context(), lockInfo, self->context->isDDEnabled(), self->ddId));
+		if (selfGeneration == self->auditInitGeneration) {
```

Comment:
Can this prevent multiple `resumeAuditStorage()`? Each `resumeAuditStorage()` is valid at the time it is being called?